### DISCUSSION TITLE
posts/Go/gin_routes_registry/

### DISCUSSION THREAD

**Discussion by Q1mi** [Q1mi/BlogComments#337] [Announcements]
https://www.liwenzhou.com/posts/Go/gin_routes_registry/ 

 李文周的Blog gin route router 路由 注册 web 框架 engine 多个 文件 目录 拆分

**Comment by veryrich** [comment]
您好，使用您这篇文章最下面的App拆分法，拆分了我的项目，但是在拆分之后，遇到了一个问题，项目启动后 Handler数量变了。

拆分之前的Handler数量一直都是4，如下图：
![img](https://s1.ax1x.com/2020/09/22/wOnMkQ.png)

拆分之后，Handler数量，变成1了，如下图：
![img](https://s1.ax1x.com/2020/09/22/wOnQYj.png)

请问发生这样的变化，会对我的项目性能，产生影响嘛？

**Comment by veryrich** [comment]
@Q1mi @qimi-blog  大佬您好~ 麻烦看一下哈~

**Comment by Q1mi** [comment]
> @Q1mi @qimi-blog 大佬您好~ 麻烦看一下哈~

是不是因为两次初始化的路由器是不一样的路由器。
我博客上一处用的是`r := gin.New()`,另外一处用的是`r := gin.Default()`,这样就会差两个中间件。

**Comment by veryrich** [comment]
感谢大佬，我做了如下更改就好了

routers 文件：更改了Init函数 加入了 r := gin.Default() 和 r.Use(Middlewares.Cors())
```go
package routers

import (
	"github.com/gin-gonic/gin"
	"sdDevOpsBackend/Middlewares"
)

type Option func(*gin.Engine)

var options []Option

// 注册app的路由配置
func Include(opts ...Option) {
	options = append(options, opts...)
}

// 初始化
func Init() *gin.Engine {
	// Gin routers
	r := gin.Default()
	r.Use(Middlewares.Cors())

	for _, opt := range options {
		opt(r)
	}
	return r
}
```

main 文件：去掉 router的定义，直接使用 routers.include，routers.Init
```go
// 加载多个APP的路由配置
routers.Include(app.Routers, department.Routers, ecs.Routers, environment.Routers, ticket.Routers, user.Routers, tempRouters)
// 初始化路由
r := routers.Init()
if err := r.Run(":8081"); err != nil {
	panic(err)
}
```
